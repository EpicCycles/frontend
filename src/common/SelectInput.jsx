import React, { Component } from 'react';
import * as PropTypes from 'prop-types';

class SelectInput extends Component {
  makeValueStrings = () => {
    const { value } = this.props;
    if (Array.isArray(value) && value.length > 0) {
      if (this.props.isMultiple) {
        return value.map(inputValue => inputValue.toString());
      } else {
        return value[0].toString();
      }
    } else if (value && value.length > 0) {
      if (this.props.isMultiple) {
        return [value.toString()];
      } else {
        return value.toString();
      }
    }
  };

  buildOptions = () => {
    const { options, isEmptyAllowed } = this.props;

    let displayOptions = [];
    if (isEmptyAllowed) {
      displayOptions.push({ value: 0, name: 'None' });
    }
    options.forEach(option => {
      const displayName = option.name ? option.name : option.value;
      displayOptions.push({
        value: option.value.toString(),
        name: displayName,
      });
    });
    return displayOptions;
  };

  findDefaultSelections = () => {
    let defaultValue = [];
    this.props.options.forEach(option => {
      if (option.isDefault) defaultValue.push(option.value.toString());
    });
    if (defaultValue.length > 0)
      if (this.props.isMultiple) {
        return defaultValue;
      } else {
        return defaultValue[0];
      }
    return 0;
  };

  handleChange = event => {
    if (this.props.isMultiple) {
      let value = [];
      const options = event.target.options;
      for (let i = 0; i < options.length; i++) {
        if (options[i].selected) {
          value.push(options[i].value);
        }
      }
      this.props.onChange(event.target.name, value);
    } else {
      this.props.onChange(event.target.name, event.target.value);
    }
  };

  render() {
    const {
      disabled,
      className,
      fieldName,
      error,
      title,
      label,
      isMultiple,
      multipleSize,
    } = this.props;
    const selectedValue = this.makeValueStrings() || this.findDefaultSelections();
    const displayOptions = this.buildOptions(selectedValue);
    let useMultipleSize = multipleSize;
    if (isMultiple && multipleSize === 1) useMultipleSize = 5;
    return (
      <div
        key={'select-container' + fieldName}
        className={className + (error ? ' error' : '')}
        title={error}
      >
        {label && <label>{label}</label>}
        <select
          name={fieldName}
          id={fieldName}
          key={fieldName}
          title={title}
          multiple={isMultiple}
          size={useMultipleSize}
          onChange={event => this.handleChange(event)}
          value={selectedValue}
          disabled={disabled}
        >
          {displayOptions.map(option => {
            return (
              <option
                id={`${fieldName}_${option.value}`}
                key={`${fieldName}_${option.value}`}
                value={option.value}
              >
                {option.name}
              </option>
            );
          })}
        </select>
      </div>
    );
  }
}

SelectInput.defaultProps = {
  className: '',
  error: '',
};
SelectInput.propTypes = {
  className: PropTypes.string,
  error: PropTypes.string,
  title: PropTypes.string,
  label: PropTypes.string,
  fieldName: PropTypes.string.isRequired,
  onChange: PropTypes.func.isRequired,
  isMultiple: PropTypes.bool,
  multipleSize: PropTypes.number,
  value: PropTypes.oneOfType([PropTypes.string, PropTypes.number]),
  isEmptyAllowed: PropTypes.bool,
  disabled: PropTypes.bool,
  options: PropTypes.array.isRequired,
};
export default SelectInput;
