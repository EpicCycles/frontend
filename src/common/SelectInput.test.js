import React from 'react';
import toJson from 'enzyme-to-json';
import SelectInput from './SelectInput';

describe('SelectInput', () => {
  test('it renders with minimum data', () => {
    const options = [{ value: 'H', name: 'Home' }];
    const component = shallow(
      <SelectInput fieldName="myField" onChange={jest.fn()} options={options} />,
    );
    expect(toJson(component)).toMatchSnapshot();
  });

  test('it renders an empty option if one is allowed', () => {
    const options = [{ value: 'H', name: 'Home' }];
    const component = shallow(
      <SelectInput fieldName="myField" onChange={jest.fn()} options={options} isEmptyAllowed />,
    );
    expect(toJson(component)).toMatchSnapshot();
    expect(component.find('option')).toHaveLength(2);
  });
  test('it renders a multiple select array if one is passed', () => {
    const options = [
      { value: 'H', name: 'Home' },
      { value: 'A', name: 'Away' },
      { value: 'O', name: 'Other' },
    ];
    const component = shallow(
      <SelectInput
        fieldName="myField"
        onChange={jest.fn()}
        options={options}
        isEmptyAllowed
        isMultiple
      />,
    );
    expect(toJson(component)).toMatchSnapshot();
    expect(component.find('option')).toHaveLength(4);
  });
  test('it renders a multiple select array if one is passed with correct size', () => {
    const options = [
      { value: 'H', name: 'Home' },
      { value: 'A', name: 'Away' },
      { value: 'O', name: 'Other' },
    ];
    const component = shallow(
      <SelectInput
        fieldName="myField"
        onChange={jest.fn()}
        options={options}
        isEmptyAllowed={false}
        isMultiple
        multipleSize={2}
      />,
    );
    expect(toJson(component)).toMatchSnapshot();
    expect(component.find('option')).toHaveLength(3);
  });
  test('it sets a value to selected when single value', () => {
    const options = [
      { value: 'H', name: 'Home' },
      { value: 'A', name: 'Away' },
      { value: 'O', name: 'Other' },
    ];
    const selectedValue = ['A'];
    const component = shallow(
      <SelectInput
        fieldName="myField"
        onChange={jest.fn()}
        options={options}
        isEmptyAllowed
        isMultiple
        value={selectedValue}
      />,
    );
    expect(toJson(component)).toMatchSnapshot();
  });
  test('it sets a value to selected when multiple select and values', () => {
    const options = [
      { value: 1, name: 'Home' },
      { value: 2, name: 'Away' },
      { value: 3, name: 'Other' },
    ];
    const selectedValue = [1, 3];
    const component = shallow(
      <SelectInput
        fieldName="myField"
        onChange={jest.fn()}
        options={options}
        isEmptyAllowed={false}
        isMultiple
        value={selectedValue}
      />,
    );
    expect(toJson(component)).toMatchSnapshot();
  });
  test('it renders the default value as selected if no value is passed', () => {
    const options = [
      { value: 'H', name: 'Home' },
      { value: 'A', name: 'Away', isDefault: true },
      { value: 'O', name: 'Other' },
    ];
    const selectedValue = [];
    const component = shallow(
      <SelectInput
        fieldName="myField"
        onChange={jest.fn()}
        options={options}
        isEmptyAllowed={false}
        isMultiple={false}
        value={selectedValue}
      />,
    );
    expect(toJson(component)).toMatchSnapshot();
  });
  test('it renders an error, title and a passed class', () => {
    const options = [
      { value: 'H', name: 'Home' },
      { value: 'A', name: 'Away', isDefault: true },
      { value: 'O', name: 'Other' },
    ];
    const selectedValue = [];
    const component = shallow(
      <SelectInput
        fieldName="myField"
        onChange={jest.fn()}
        options={options}
        isEmptyAllowed={false}
        isMultiple={false}
        value={selectedValue}
        error="show an error"
        title="show a title"
        className="pink"
      />,
    );
    expect(toJson(component)).toMatchSnapshot();
  });
  test('it shows the only choice as selected when the array has length 1', () => {
    const options = [{ value: 'H', name: 'Home' }];
    const component = shallow(
      <SelectInput fieldName="myField" onChange={jest.fn()} options={options} isEmptyAllowed />,
    );
    expect(component.find('select').prop('value')).toEqual('H');
  });
});
