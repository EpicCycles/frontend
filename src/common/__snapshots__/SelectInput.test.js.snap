// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SelectInput it renders a multiple select array if one is passed 1`] = `
<div
  className=""
  key="select-containermyField"
  title=""
>
  <select
    id="myField"
    key="myField"
    multiple={true}
    name="myField"
    onChange={[Function]}
    value={Array []}
  >
    <option
      id="myField_0"
      key="myField_0"
      value={0}
    >
      None
    </option>
    <option
      id="myField_H"
      key="myField_H"
      value="H"
    >
      Home
    </option>
    <option
      id="myField_A"
      key="myField_A"
      value="A"
    >
      Away
    </option>
    <option
      id="myField_O"
      key="myField_O"
      value="O"
    >
      Other
    </option>
  </select>
</div>
`;

exports[`SelectInput it renders a multiple select array if one is passed with correct size 1`] = `
<div
  className=""
  key="select-containermyField"
  title=""
>
  <select
    id="myField"
    key="myField"
    multiple={true}
    name="myField"
    onChange={[Function]}
    size={2}
    value={Array []}
  >
    <option
      id="myField_H"
      key="myField_H"
      value="H"
    >
      Home
    </option>
    <option
      id="myField_A"
      key="myField_A"
      value="A"
    >
      Away
    </option>
    <option
      id="myField_O"
      key="myField_O"
      value="O"
    >
      Other
    </option>
  </select>
</div>
`;

exports[`SelectInput it renders an empty option if one is allowed 1`] = `
<div
  className=""
  key="select-containermyField"
  title=""
>
  <select
    id="myField"
    key="myField"
    name="myField"
    onChange={[Function]}
    value={0}
  >
    <option
      id="myField_0"
      key="myField_0"
      value={0}
    >
      None
    </option>
    <option
      id="myField_H"
      key="myField_H"
      value="H"
    >
      Home
    </option>
  </select>
</div>
`;

exports[`SelectInput it renders an error, title and a passed class 1`] = `
<div
  className="pink error"
  key="select-containermyField"
  title="show an error"
>
  <select
    id="myField"
    key="myField"
    multiple={false}
    name="myField"
    onChange={[Function]}
    title="show a title"
    value="A"
  >
    <option
      id="myField_H"
      key="myField_H"
      value="H"
    >
      Home
    </option>
    <option
      id="myField_A"
      key="myField_A"
      value="A"
    >
      Away
    </option>
    <option
      id="myField_O"
      key="myField_O"
      value="O"
    >
      Other
    </option>
  </select>
</div>
`;

exports[`SelectInput it renders the default value as selected if no value is passed 1`] = `
<div
  className=""
  key="select-containermyField"
  title=""
>
  <select
    id="myField"
    key="myField"
    multiple={false}
    name="myField"
    onChange={[Function]}
    value="A"
  >
    <option
      id="myField_H"
      key="myField_H"
      value="H"
    >
      Home
    </option>
    <option
      id="myField_A"
      key="myField_A"
      value="A"
    >
      Away
    </option>
    <option
      id="myField_O"
      key="myField_O"
      value="O"
    >
      Other
    </option>
  </select>
</div>
`;

exports[`SelectInput it renders with minimum data 1`] = `
<div
  className=""
  key="select-containermyField"
  title=""
>
  <select
    id="myField"
    key="myField"
    name="myField"
    onChange={[Function]}
    value={0}
  >
    <option
      id="myField_H"
      key="myField_H"
      value="H"
    >
      Home
    </option>
  </select>
</div>
`;

exports[`SelectInput it sets a value to selected when multiple select and values 1`] = `
<div
  className=""
  key="select-containermyField"
  title=""
>
  <select
    id="myField"
    key="myField"
    multiple={true}
    name="myField"
    onChange={[Function]}
    value={
      Array [
        "1",
        "3",
      ]
    }
  >
    <option
      id="myField_1"
      key="myField_1"
      value="1"
    >
      Home
    </option>
    <option
      id="myField_2"
      key="myField_2"
      value="2"
    >
      Away
    </option>
    <option
      id="myField_3"
      key="myField_3"
      value="3"
    >
      Other
    </option>
  </select>
</div>
`;

exports[`SelectInput it sets a value to selected when single value 1`] = `
<div
  className=""
  key="select-containermyField"
  title=""
>
  <select
    id="myField"
    key="myField"
    multiple={true}
    name="myField"
    onChange={[Function]}
    value={
      Array [
        "A",
      ]
    }
  >
    <option
      id="myField_0"
      key="myField_0"
      value={0}
    >
      None
    </option>
    <option
      id="myField_H"
      key="myField_H"
      value="H"
    >
      Home
    </option>
    <option
      id="myField_A"
      key="myField_A"
      value="A"
    >
      Away
    </option>
    <option
      id="myField_O"
      key="myField_O"
      value="O"
    >
      Other
    </option>
  </select>
</div>
`;
